### PR TITLE
Munchausen RL

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
Implementation of the Munchausen RL.

Tested on Breakout.   The reward-weight is an import hyper-parameter for tuning.
Currently, it seems that with a proper weight (e.g. 0.5), there are some gain initially, but eventually similar to SAC.
image

### CODE REVIEW

**Comment by emailweixu** [comment]
Are there any experiment results?

File: alf/algorithms/sac_algorithm.py [hunk 1]
```diff
@@ -846,6 +857,30 @@ def _calc_critic_loss(self, info: SacInfo):
...
+                    log_pi_rollout_a = sum(nest.flatten(log_pi_rollout_a))
+
+                munchausen_reward = nest.map_structure(
+                    lambda la, lp: torch.exp(la) * lp, self._log_alpha,
+                    log_pi_rollout_a)
+                munchausen_reward = sum(nest.flatten(munchausen_reward))
+                info = info._replace(
+                    reward=(
+                        info.reward + self._munchausen_reward_weight *
+                        common.expand_dims_as(munchausen_reward, info.reward)))
```
Comment (hnyu) [inline alf/algorithms/sac_algorithm.py:905]: This reward should be added to the next time step? Because you are using the rollout action of the current step and its reward is contained by the next step. The entropy reward and this reward shouldn't be added to the same step.

**Comment by Haichao-Zhang** [comment]
> Are there any experiment results?

When submitting the PR, only tested on pendulum for sanity check. On pendulum, no clear difference compared to SAC.
I'm running more experiments. Will post some here when available.

File: alf/algorithms/sac_algorithm.py [hunk 2]
```diff
@@ -846,6 +857,30 @@ def _calc_critic_loss(self, info: SacInfo):
         When the reward is multi-dim, the entropy reward will be added to *all*
         dims.
         """
```
Comment (hnyu) [inline alf/algorithms/sac_algorithm.py:856]: The comment above has been separated from the entropy reward code.

File: alf/algorithms/sac_algorithm.py (hunk 1 above)
Comment (Haichao-Zhang) [reply alf/algorithms/sac_algorithm.py:905]: This is a bug. Fixed

File: alf/algorithms/sac_algorithm.py (hunk 2 above)
Comment (Haichao-Zhang) [reply alf/algorithms/sac_algorithm.py:856]: Now make the comment embedded to the right place.

File: alf/algorithms/sac_algorithm.py [hunk 3]
```diff
@@ -856,10 +864,46 @@ def _calc_critic_loss(self, info: SacInfo):
...
+                    lambda dist, a: dist.log_prob(a), info.action_distribution,
+                    info.action)
+
+                if self._act_type == ActionType.Mixed:
+                    # For mixed type, add log_pi separately
+                    log_pi_rollout_a = type(self._action_spec)(
+                        (sum(nest.flatten(log_pi_rollout_a[0])),
+                         sum(nest.flatten(log_pi_rollout_a[1]))))
+                else:
+                    log_pi_rollout_a = sum(nest.flatten(log_pi_rollout_a))
```
Comment (hnyu) [inline alf/algorithms/sac_algorithm.py:886]: Why do you want to first add log_pi here?

Comment (hnyu) [inline alf/algorithms/sac_algorithm.py:886]: Why do you want to first add log_pi here? (Unless for each log_alpha you have multiple log_pis)

File: alf/algorithms/sac_algorithm.py [hunk 4]
```diff
@@ -267,6 +273,11 @@ def __init__(self,
             critic_network_cls, q_network_cls)
 
         self._use_entropy_reward = use_entropy_reward
+        self._munchausen_reward_weight = max(0, munchausen_reward_weight)
+        if munchausen_reward_weight > 0:
+            assert not normalize_entropy_reward, (
```
Comment (hnyu) [inline alf/algorithms/sac_algorithm.py:278]: normalize_entropy_reward has no conflict with munchausen_reward_weight > 0?

Comment (hnyu) [inline alf/algorithms/sac_algorithm.py:278]: normalize_entropy_reward has no conflict with munchausen_reward_weight > 0?

**Review by hnyu — CHANGES_REQUESTED** [review]
